### PR TITLE
Maemo/stretch

### PR DESCRIPTION
**Pull request by spinal84**
* Add systemd support
* Tidy up

### CODE REVIEW

File: mce-hal.c [hunk 1]
```diff
@@ -55,27 +55,29 @@ product_id_t get_product_id(void)
...
+	 * so must be checked after
+	 */
 	if (mce_read_string_from_file(CPUINFO_PATH, &tmp)) {
 		if (strstr(tmp, "Nokia 770") != NULL) {
 			product_id = PRODUCT_SU18;
 		} else if (strstr(tmp, "Nokia N800") != NULL) {
 			product_id = PRODUCT_RX34;
 		} else if (strstr(tmp, "Nokia N810 WiMAX") != NULL) {
 			product_id = PRODUCT_RX48;
-		} else if (strstr(tmp, "Nokia N810") != NULL) { /* NOTE: "Nokia N810" is substring of "Nokia N810 WiMAX" so must be checked after */
```
Comment (MerlijnWajer) [inline mce-hal.c:68]: Why is this removed?

Comment (MerlijnWajer) [reply mce-hal.c:68]: Or rather, moved.

File: mce-hal.c [hunk 2]
```diff
@@ -38,7 +38,7 @@ product_id_t get_product_id(void)
 	gchar *tmp = NULL;
 
 	if (product_id != PRODUCT_UNSET)
-		goto EXIT;
```
Comment (MerlijnWajer) [inline mce-hal.c:41]: Does this fix any compiler warnings or errors? Or did you want to get rid of the goto?

File: modules/homekey.c [hunk 3]
```diff
@@ -173,7 +173,7 @@ const gchar *g_module_check_init(GModule *module)
 	(void)module;
 
 	if (get_product_id() == PRODUCT_RX51)
-		goto EXIT;
```
Comment (MerlijnWajer) [inline modules/homekey.c:176]: Did you just want to get rid of the goto?

File: mce-hal.c (hunk 1 above)
Comment (spinal84) [reply mce-hal.c:68]: This comment made the line to be 130+ chars long.
After I moved it to another (appropriate) place, this line doesn't exceed 80 chars now.
Which I think is good. Don't you?

File: mce-hal.c (hunk 2 above)
Comment (spinal84) [reply mce-hal.c:41]: Excessive goto's in high level programming languages are considered harmful.
It's better to avoid using goto without a reason.

It's not a taboo. I'm using goto's when it makes sense in the code to make it more simple and/or easier to read. But this is not the case here.

Some references:
https://github.com/maemo-leste/status-area-applet-battery/commit/761aef615286241a32c07a765e221e63dcc0fdc3
http://www.cs.utexas.edu/users/EWD/ewd02xx/EWD215.PDF
https://stackoverflow.com/questions/3517726/what-is-wrong-with-using-goto

File: mce-hal.c (hunk 1 above)
Comment (MerlijnWajer) [reply mce-hal.c:68]: I think the comment would be better in the context, so maybe above the `} else if (strstr(tmp, "Nokia N810") != NULL)` line.

File: mce-hal.c (hunk 2 above)
Comment (MerlijnWajer) [reply mce-hal.c:41]: Sure, but I am not sure if it made sense to remove them here. I don't care much either way, but when I modify functions to add functionality, I try to stick to how the function originally worked, unless it's clearly broken. :p 

In this case, nothing was added to this function, as far as I can tell. You just moved some comments and goto into return - as opposed to the other commits that fix compiler warnings. :)

**Comment by MerlijnWajer** [comment]
Overall, looks good to me. @freemangordon - got an opinion?

File: mce-hal.c (hunk 2 above)
Comment (spinal84) [reply mce-hal.c:41]: That's why I called the commit "Minor fixes": https://github.com/maemo-leste/mce/pull/3/commits/065823f5d8773ea6ae080c0d5958d356dad73855
Sorry, if I didn't get the idea from your last message correct. Could you please explain?

Comment (MerlijnWajer) [reply mce-hal.c:41]: On the face, nothing was particularly wrong with this function. It had no logic mistakes, and the comment was maybe on a line that was too long, but it was in context. I would have probably just left it as is, and let `indent` at some point take care of it.

But it's fine this way, if you move the comment more in context.

Comment (MerlijnWajer) [reply mce-hal.c:41]: Sorry, the comment part was part of the change above, not here.

File: mce-hal.c (hunk 1 above)
Comment (spinal84) [reply mce-hal.c:68]: For me it will become uggly. Compare.
1) Now:
```
        /*
         * NOTE: "Nokia N810" is substring of "Nokia N810 WiMAX",
         * so must be checked after
         */
        if (mce_read_string_from_file(CPUINFO_PATH, &tmp)) {
                if (strstr(tmp, "Nokia 770") != NULL) {
                        product_id = PRODUCT_SU18;
                } else if (strstr(tmp, "Nokia N800") != NULL) {
                        product_id = PRODUCT_RX34;
                } else if (strstr(tmp, "Nokia N810 WiMAX") != NULL) {
                        product_id = PRODUCT_RX48;
                } else if (strstr(tmp, "Nokia N810") != NULL) {
                        product_id = PRODUCT_RX44;
                } else if (strstr(tmp, "Nokia RX-51 board") != NULL) {
                        product_id = PRODUCT_RX51;
                } else {
                        product_id = PRODUCT_UNKNOWN;
                }
                g_free(tmp);
        }

```
2) After the change (vim auto indentation):
```
        if (mce_read_string_from_file(CPUINFO_PATH, &tmp)) {
                if (strstr(tmp, "Nokia 770") != NULL) {
                        product_id = PRODUCT_SU18;
                } else if (strstr(tmp, "Nokia N800") != NULL) {
                        product_id = PRODUCT_RX34;
                } else if (strstr(tmp, "Nokia N810 WiMAX") != NULL) {
                        product_id = PRODUCT_RX48;
                        /*
                         * NOTE: "Nokia N810" is substring of "Nokia N810 WiMAX",
                         * so must be checked after
                         */
                } else if (strstr(tmp, "Nokia N810") != NULL) {
                        product_id = PRODUCT_RX44;
                } else if (strstr(tmp, "Nokia RX-51 board") != NULL) {
                        product_id = PRODUCT_RX51;
                } else {
                        product_id = PRODUCT_UNKNOWN;
                }
                g_free(tmp);
        }
```

3) After the change (manual indentation):
```
        if (mce_read_string_from_file(CPUINFO_PATH, &tmp)) {
                if (strstr(tmp, "Nokia 770") != NULL) {
                        product_id = PRODUCT_SU18;
                } else if (strstr(tmp, "Nokia N800") != NULL) {
                        product_id = PRODUCT_RX34;
                } else if (strstr(tmp, "Nokia N810 WiMAX") != NULL) {
                        product_id = PRODUCT_RX48;
                /*      
                 * NOTE: "Nokia N810" is substring of "Nokia N810 WiMAX",
                 * so must be checked after
                 */
                } else if (strstr(tmp, "Nokia N810") != NULL) {
                        product_id = PRODUCT_RX44;
                } else if (strstr(tmp, "Nokia RX-51 board") != NULL) {
                        product_id = PRODUCT_RX51;
                } else {
                        product_id = PRODUCT_UNKNOWN;
                }
                g_free(tmp);
        }
```

Comment (spinal84) [reply mce-hal.c:68]: Do you still want this change?

Comment (MerlijnWajer) [reply mce-hal.c:68]: I think the manual indentation looks good. If this if statement ever grows to >50+ devices, it's also better to have the comment closer to the line it relates to.

Comment (MerlijnWajer) [reply mce-hal.c:68]: Thanks!

Comment (spinal84) [reply mce-hal.c:68]: Such variant will have auto indentation in place and the comment is in appropriate context:
```
        if (mce_read_string_from_file(CPUINFO_PATH, &tmp)) {
                if (strstr(tmp, "Nokia 770") != NULL) {
                        product_id = PRODUCT_SU18;
                } else if (strstr(tmp, "Nokia N800") != NULL) {
                        product_id = PRODUCT_RX34;
                } else if (strstr(tmp, "Nokia N810 WiMAX") != NULL) {
                        product_id = PRODUCT_RX48;
                } else if (strstr(tmp, "Nokia N810") != NULL) {
                        /*
                         * NOTE: "Nokia N810" is substring of "Nokia N810
                         * WiMAX", so must be checked after
                         */
                        product_id = PRODUCT_RX44;
                } else if (strstr(tmp, "Nokia RX-51 board") != NULL) {
                        product_id = PRODUCT_RX51;
                } else {
                        product_id = PRODUCT_UNKNOWN;
                }
                g_free(tmp);
        }

```
What do you think about it?

Comment (spinal84) [reply mce-hal.c:68]: Maybe just remove this comment?
I doubt someone will change the order of these if's in the future.

Comment (MerlijnWajer) [reply mce-hal.c:68]: I am fine with the manual indentation or the above auto indentation, let's not remove it.

File: mce-hal.c (hunk 2 above)
Comment (spinal84) [reply mce-hal.c]: > Sorry, the comment part was part of the change above, not here.

AFAIK, you can just remove the wrong comment from github.

Comment (spinal84) [reply mce-hal.c]: test

File: mce-hal.c (hunk 1 above)
Comment (spinal84) [reply mce-hal.c:68]: Fixed. Please check if you are satisfied with how it's done now.

Comment (MerlijnWajer) [reply mce-hal.c:68]: Comment looks good.

File: debian/mce.init [hunk 4]
```diff
@@ -1,4 +1,14 @@
...
+### BEGIN INIT INFO
+# Provides:          mce
+# Required-Start:    $remote_fs dsme dbus dbus-user
+# Required-Stop:     $remote_fs dsme dbus dbus-user
+# Default-Start:     2 3 4 5
+# Default-Stop:      0 1 6
+# Short-Description: Mode Control Entity
+# Description:       This init script starts the mode control software
+#                    used on the Maemo platform for Internet Tablets
+### END INIT INFO
```
Comment (MerlijnWajer) [inline debian/mce.init]: So I did this for one or two packages because @parazyd asked me to do it, but it seems to break even just running the scripts with openrc. I am not sure if you tested this, or are just adding it because I added to a few packages, but probably don't want to spend a lot more time doing this right now.

Comment (spinal84) [reply debian/mce.init]: I do this because missing LSB headers break systemd setup as well.
It doesn't allow to enable/disable service because of errors like this:
`update-rc.d: error: mce Default-Start contains no runlevels, aborting.`

Comment (spinal84) [reply debian/mce.init]: > So I did this for one or two packages because @parazyd asked me to do it, but it seems to break even just running the scripts with openrc. I am not sure if you tested this, or are just adding it because I added to a few packages, but probably don't want to spend a lot more time doing this right now.

How to reproduce the errors you get ("it seems to break even just running the scripts with openrc")?

Comment (MerlijnWajer) [reply debian/mce.init]: Let's move this discussion to an issue/mail/irc, since we'll definitely want to record this somewhere, and also get Parazyd's input. I need to go out for ~1 hour or so, but will be around after that.

In my case, running `/etc/init.d/droid4-powermanagement status` didn't work when the lsb headers were present. Removing them made it work. Perhaps `runscript` should be used rather than `openrc-run`.

Comment (MerlijnWajer) [reply debian/mce.init]: I am OK with the PR, apart from these LSB headers. This currently will break builds. So please remove it for now, and let's add it later when we're all ready for this.

Comment (spinal84) [reply debian/mce.init]: > I am OK with the PR, apart from these LSB headers. This currently will break builds. So please remove it for now, and let's add it later when we're all ready for this.

Done.

Comment (spinal84) [reply debian/mce.init]: BTW, I already debugged openrc and know how to fix LSB headers not working with it.
All you need is to comment or remove 321 - 322 lines from /lib/rc/sh/openrc-run.sh

Comment (MerlijnWajer) [reply debian/mce.init]: cc @parazyd ^^

**Comment by MerlijnWajer** [comment]
FWIW, this looks good to me. I did merge some other pull requests in between, so I want to make sure that the debian/changelog at least contains those entries too. Maybe once I'm done with ALS and touch disable (should be tomorrow - Friday), we can merge all of the open PRs.

**Comment by MerlijnWajer** [comment]
I will look at/merge this once I've done another release of mce with my current droid4 fixes; so we might need the changelog changed.

**Comment by MerlijnWajer** [comment]
Merged it anyway, thanks!
